subcommand/get-consul-client-ca/command.go

```diff
@@ -0,0 +1,262 @@
+			return err
+		}
+
+		return nil
+	}, backoff.NewConstantBackOff(1*time.Second))
+
+	err = ioutil.WriteFile(c.flagOutputFile, []byte(activeRoot), 0644)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error writing CA file: %s", err))
+		return 1
+	}
+
+	c.UI.Info(fmt.Sprintf("Successfully wrote Consul client CA to: %s", c.flagOutputFile))
+	return 0
+}
+
+// consulClient returns a Consul API client.
+func (c *Command) consulClient(logger hclog.Logger) (*api.Client, error) {
+	// Create default Consul config.
+	// This will also read any environment variables.
+	cfg := api.DefaultConfig()
+
+	// change the scheme to HTTPS
+	// since we don't want to send unencrypted requests
+	cfg.Scheme = "https"
+
+	addr, err := c.consulServerAddr(logger)
+	if err != nil {
+		return nil, err
+	}
+	if addr != "" {
+		cfg.Address = addr
+	}
+
+	// Set the CA file and TLS server name if the flag is provided.
+	// This will overwrite any env variables values for these flags.
+	if c.flagCAFile != "" {
+		cfg.TLSConfig.CAFile = c.flagCAFile
+	}
+	if c.flagTLSServerName != "" {
+		cfg.TLSConfig.Address = c.flagTLSServerName
+	}
+
+	return api.NewClient(cfg)
+}
+
+// consulServerAddr returns the consul server address
+// as a string in the <server_ip_or_dns_name>:<server_port> format.
+//
+// 1. If the server address is a cloud auto-join URL,
+//    it calls go-discover library to discover server addresses,
+//    picks the first address from the list and uses the provided port.
+// 2. Otherwise, it uses the address provided by the -server-addr
+//    and the -server-port flags.
+func (c *Command) consulServerAddr(logger hclog.Logger) (string, error) {
+	// First, check if the server address is a cloud auto-join string.
+	// If it is, discover server addresses through the cloud provider.
+	// This code was adapted from
+	// https://github.com/hashicorp/consul/blob/c5fe112e59f6e8b03159ec8f2dbe7f4a026ce823/agent/retry_join.go#L55-L89.
+	if strings.Contains(c.flagServerAddr, "provider=") {
+		disco, err := c.newDiscover()
+		if err != nil {
+			return "", err
+		}
+		logger.Debug("using cloud auto-join", "server-addr", c.flagServerAddr)
+		servers, err := disco.Addrs(c.flagServerAddr, logger.StandardLogger(&hclog.StandardLoggerOptions{
+			InferLevels: true,
+		}))
+		if err != nil {
+			return "", err
+		}
+
+		// check if we discovered any servers
+		if len(servers) == 0 {
+			return "", fmt.Errorf("could not discover any Consul servers with %q", c.flagServerAddr)
+		}
+
+		logger.Debug("discovered servers", strings.Join(servers, " "))
+
+		// Pick the first server from the list,
+		// ignoring the port since we need to use HTTP API
+		// and don't care about the RPC port.
+		firstServer := strings.SplitN(servers[0], ":", 2)[0]
+		return fmt.Sprintf("%s:%s", firstServer, c.flagServerPort), nil
+	}
+
+	// Otherwise, return serverAddr:serverPort set by the provided flags
+	return fmt.Sprintf("%s:%s", c.flagServerAddr, c.flagServerPort), nil
```

Comment:
I'd put this case first since it's shorter. i.e. `if !strings.contains(provider) { ... }`